Logging solution replacement for oklog with grafana/loki (#2514)
kyma-project/kyma: docs/logging/docs/02-01-logging.md

title: Architecture
---

> This document provides an overview of the logging architecture in Kyma. It describes the information sources from which promtail extracts logs to feed to Loki.

Reviewer: ```suggestion
This document provides an overview of the logging architecture in Kyma. 
```
I would remove the second part, since it's rather about the agent itself, log chunks and the read/write process - or we can add one line about them.

## Agent (Promtail)

> Promtail is the agent responsible for collecting reliable metadata, consistent with the time series or metrics metadata. To achieve this, the agent uses the same service discovery and relabelling libraries as Prometheus. It is used as a Deamon Set to discover targets, create metadata labels, and tail log files to produce a stream of logs. The logs are buffered on the client side and then sent to the service.

Reviewer: ```suggestion
Promtail is the agent responsible for collecting reliable metadata, consistent with the time series or metrics metadata. To achieve this, the agent uses the same service discovery and relabelling libraries as Prometheus. Promtail is used as a Deamon Set to discover targets, create metadata labels, and tail log files to produce a stream of logs. The logs are buffered on the client side and then sent to the service.
```

#### Life of a Write Request
The server-side components on the write path wii mirror the [Cortex](https://github.com/cortexproject/cortex) architecture.
* Writes will first hit the Distributor, which is responsible for distributing and replacing the writes to the ingesters. Loki use the Cortex consistent hash ring and distribute writes based on a hash of the entire metadata.
* Next writes will hit a 'log ingester' which batches up writes for the same stream in memory in to 'log chunks'. When chunks reach a predefined size or age, periodically flushed to the Cortex chunk store.

> * The Cortex chunk store will be updated to reduce copying of chunk data on the read and write path and add support for writing chunks of Grafana.

Author: doc updated, please review again

#### Log Chunks

> A chunk is all logs for a given label set over a certain period. the chunks support appends, seek and streaming reads.

Reviewer: ```suggestion
A log chunk consists of all logs for a given label, collected over a certain time period. Log chunks support append, seek, and stream operations on read requests.
```
For the second part - the append, seek and stream are the types of operations on read requests? I tried researching this topic a bit, but found just the info on operations such as seek, append or stream.

Author: No actually append + seek is write operation on a chunk, seek + stream is a read operation on a chunk

## Life of a write request
The server-side components on the write path will mirror the [Cortex](https://github.com/cortexproject/cortex) architecture.
* Writes will first hit the Distributor, which is responsible for distributing and replacing the writes to the ingesters. Loki use the Cortex consistent hash ring and distribute writes based on a hash of the entire metadata.
* Next writes will hit a 'log ingester' which batches up writes for the same stream in memory in to 'log chunks'. When chunks reach a predefined size or age, periodically flushed to the Cortex chunk store.
* The Cortex chunk store will be updated to reduce copying of chunk data on the read and write path and add support for writing chunks of google cloud storage.

#### Log Chunks
A log chunk consists of all logs for a given metadata (e.g. labels), collected over a certain time period. Log chunks support append, seek, and stream operations on read requests.

#### Life of a Query Request

> As chunks are larger than Prometheus/Cortex chunks (Cortex chunks are max 1KB in size), it is not possible to load and decompress them in their entirety. 

Author: We can't say Prometheus or Cortex because Cortex is a storage engine which prometheus use to store chunks like loki. What I want to highlight here was "Prometheus Cortex chunks are max 1K but loki Cortex chunks can be larger"

> As chunks are larger than Prometheus/Cortex chunks (Cortex chunks are max 1KB in size), it is not possible to load and decompress them in their entirety. 

Reviewer 2: I changed it to Prometheus-Cortex .
